hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/sink/StreamWriteOperatorCoordinator.java

```diff
@@ -501,44 +503,22 @@ private void handleWriteMetaEvent(WriteMetadataEvent event) {
     addEventToBuffer(event);
   }
 
-  /**
-   * The coordinator reuses the instant if there is no data for this round of checkpoint,
-   * sends the commit ack events to unblock the flushing.
-   */
-  private void sendCommitAckEvents(long checkpointId) {
-    CompletableFuture<?>[] futures = Arrays.stream(this.gateways).filter(Objects::nonNull)
-        .map(gw -> gw.sendEvent(CommitAckEvent.getInstance(checkpointId)))
-        .toArray(CompletableFuture<?>[]::new);
-    CompletableFuture.allOf(futures).whenComplete((resp, error) -> {
-      if (!sendToFinishedTasks(error)) {
-        throw new HoodieException("Error while waiting for the commit ack events to finish sending", error);
-      }
-    });
-  }
-
-  /**
-   * Decides whether the given exception is caused by sending events to FINISHED tasks.
-   *
-   * <p>Ugly impl: the exception may change in the future.
-   */
-  private static boolean sendToFinishedTasks(Throwable throwable) {
-    return throwable.getCause() instanceof TaskNotRunningException
-        || throwable.getCause().getMessage().contains("running");
-  }
-
   /**
    * Commits the instant.
    */
-  private boolean commitInstant(String instant) {
-    return commitInstant(instant, -1);
+  private boolean commitInstants(long checkpointId) {
+    // use < instead of <= because the write metadata event sends the last known checkpoint id which is smaller than the current one.
+    List<Boolean> result = this.eventBuffers.entrySet().stream().filter(entry -> entry.getKey() < checkpointId)
+        .map(entry -> commitInstant(entry.getKey(), entry.getValue().getLeft(), entry.getValue().getRight())).collect(Collectors.toList());
```

Comment:
It looks like a one-time commit of all instants before. Here need to ensure that all historical commits are complete and intact.

Reply:
See the cleaing of bootstrap event and the strategy for commiting empty instant.